pkg/apiserver/watch.go

```diff
@@ -64,119 +59,218 @@ func (w *realTimeoutFactory) TimeoutCh() (<-chan time.Time, func() bool) {
+
+	t timeoutFactory
 }
 
-// ServeHTTP serves a series of JSON encoded events via straight HTTP with
-// Transfer-Encoding: chunked.
-func (self *WatchServer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
-	loggedW := httplog.LogOf(req, w)
+// Serve serves a series of encoded events via HTTP with Transfer-Encoding: chunked
+// or over a websocket connection.
+func (s *WatchServer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 	w = httplog.Unlogged(w)
-	timeoutCh, cleanup := self.t.TimeoutCh()
-	defer cleanup()
-	defer self.watching.Stop()
+
+	if wsstream.IsWebSocketRequest(req) {
+		w.Header().Set("Content-Type", s.mediaType)
+		websocket.Handler(s.HandleWS).ServeHTTP(w, req)
+		return
+	}
 
 	cn, ok := w.(http.CloseNotifier)
 	if !ok {
-		loggedW.Addf("unable to get CloseNotifier: %#v", w)
-		http.NotFound(w, req)
+		err := fmt.Errorf("unable to start watch - can't get http.CloseNotifier: %#v", w)
+		utilruntime.HandleError(err)
+		s.scope.err(errors.NewInternalError(err), w, req)
 		return
 	}
 	flusher, ok := w.(http.Flusher)
 	if !ok {
-		loggedW.Addf("unable to get Flusher: %#v", w)
-		http.NotFound(w, req)
+		err := fmt.Errorf("unable to start watch - can't get http.Flusher: %#v", w)
+		utilruntime.HandleError(err)
+		s.scope.err(errors.NewInternalError(err), w, req)
+		return
+	}
+
+	// get a framed encoder
+	f, ok := s.encoder.(streaming.Framer)
+	if !ok {
+		// programmer error
+		err := fmt.Errorf("no streaming support is available for media type %q", s.mediaType)
+		utilruntime.HandleError(err)
+		s.scope.err(errors.NewBadRequest(err.Error()), w, req)
+		return
+	}
+	framer := f.NewFrameWriter(w)
+	if framer == nil {
+		// programmer error
+		err := fmt.Errorf("no stream framing support is available for media type %q", s.mediaType)
+		utilruntime.HandleError(err)
+		s.scope.err(errors.NewBadRequest(err.Error()), w, req)
 		return
 	}
+	e := streaming.NewEncoder(framer, s.encoder)
+
+	// ensure the connection times out
+	timeoutCh, cleanup := s.t.TimeoutCh()
+	defer cleanup()
+	defer s.watching.Stop()
+
+	// begin the stream
+	w.Header().Set("Content-Type", s.mediaType)
 	w.Header().Set("Transfer-Encoding", "chunked")
 	w.WriteHeader(http.StatusOK)
 	flusher.Flush()
-	// TODO: use arbitrary serialization on watch
-	encoder := watchjson.NewEncoder(w, self.encoder)
+
+	buf := &bytes.Buffer{}
 	for {
 		select {
 		case <-cn.CloseNotify():
 			return
 		case <-timeoutCh:
 			return
-		case event, ok := <-self.watching.ResultChan():
+		case event, ok := <-s.watching.ResultChan():
 			if !ok {
 				// End of results.
 				return
 			}
-			self.fixup(event.Object)
-			if err := encoder.Encode(&event); err != nil {
-				// Client disconnect.
+			obj := event.Object
+			s.fixup(obj)
+			if err := s.embeddedEncoder.EncodeToStream(obj, buf); err != nil {
+				// unexpected error
+				utilruntime.HandleError(fmt.Errorf("unable to encode watch object: %v", err))
+				return
+			}
+			event.Object = &runtime.Unknown{
```

Comment:
Sorry - I'm probably missing something. So the "Encode" method in case of runtime.Unknown is not serializating the Unknown object, but is just taking the Raw field and returns it? Is that the case?